ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl/Pure.hs

```diff
@@ -0,0 +1,321 @@
+-- them.
+data RemoveTxs blk = WriteRemoveTxs (InternalState blk) (Maybe (TraceEventMempool blk))
+
+-- | Intepret a 'RemoveTxs' with the resulting values produced by manually
+-- removing the transactions given to 'pureRemovTxs' from the mempool.
+runRemoveTxs :: forall m blk. IOLike m
+  => StrictTVar m (InternalState blk)
+  -> RemoveTxs blk
+  -> STM m (Maybe (TraceEventMempool blk))
+runRemoveTxs stateVar (WriteRemoveTxs is t) = do
+  writeTVar stateVar is
+  return t
+
+-- | Craft a 'RemoveTxs' that manually removes the given transactions from the
+-- mempool, returning inside it an updated InternalState.
+pureRemoveTxs
+  :: ( LedgerSupportsMempool blk
+     , HasTxId (GenTx blk)
+     , ValidateEnvelope blk
+     )
+  => LedgerConfig blk
+  -> MempoolCapacityBytesOverride
+  -> [GenTxId blk]
+  -> InternalState blk
+  -> LedgerState blk
+  -> RemoveTxs blk
+pureRemoveTxs cfg capacityOverride txIds IS { isTxs, isLastTicketNo } lstate =
+  -- Filtering is O(n), but this function will rarely be used, as it is an
+  -- escape hatch when there's an inconsistency between the ledger and the
+  -- mempool.
+  let toRemove = Set.fromList txIds
+      txTickets' = filter
+                   ((`notElem` toRemove) . txId . txForgetValidated . txTicketTx)
+                   (TxSeq.toList isTxs)
+      (slot, ticked) = tickLedgerState cfg (ForgeInUnknownSlot lstate)
+      vr = revalidateTxsFor
+            capacityOverride
+            cfg
+            slot
+            ticked
+            isLastTicketNo
+            txTickets'
+      is' = internalStateFromVR vr
+      needsTrace =
+        if null txIds
+        then Nothing
+        else Just $
+               TraceMempoolManuallyRemovedTxs txIds (map fst (vrInvalid vr)) (isMempoolSize is')
+  in WriteRemoveTxs is' needsTrace
+
+-- | A datatype containing the new state produced by syncing with the Ledger, a
+-- snapshot of that mempool state and, if needed, a tracing message.
+data SyncWithLedger blk = NewSyncedState (InternalState blk) (MempoolSnapshot blk TicketNo) (Maybe (TraceEventMempool blk))
+
+-- | Intepret a 'SyncWithLedger' value produced by syncing the transactions in
+--  the mempool with the current ledger state of the 'ChainDB'.
+--
+-- The transactions that exist within the mempool will be revalidated
+-- against the current ledger state. Transactions which are found to be
+-- invalid with respect to the current ledger state, will be dropped
+-- from the mempool, whereas valid transactions will remain.
+--
+-- n.b. in our current implementation, when one opens a mempool, we
+-- spawn a thread which performs this action whenever the 'ChainDB' tip
+-- point changes.
+runSyncWithLedger :: forall m blk. IOLike m
+  => StrictTVar m (InternalState blk)
+  -> SyncWithLedger blk
+  -> STM m (Maybe (TraceEventMempool blk), MempoolSnapshot blk TicketNo)
+runSyncWithLedger stateVar (NewSyncedState is msp mTrace) = do
```

Comment:
Style: generally we indent by 4 even if there's no where clause, because there might be a where clause later.

But since you're consistent in the module, this is probably fine.

Reply:
More like my tooling is consistent in indenting with 2 despite I having set
```
(haskell-indent-offset 4)
```
Fixed this one (and some others). Resolving